### DISCUSSION TITLE
Can we support using symbols as keys?

### DISCUSSION THREAD

**Discussion by jackmellis** [TanStack/query#477] [General] [answered]
I was really hoping that moving away from redux meant I could stop having awful action type strings like `@@foo/bah/SOME_ACTION` all over the place.

I still need to pass a key in to queries to ensure they're unique so now I have `@@foo/bah/SOME_QUERY_KEY` everywhere. It'd be much nicer if I could _finally_ find a place to use symbols!

```ts
const Basket = Symbol('basket')

const [ basket ] = useQuery(Basket, fetchBasket)
```

I guess the biggest complication is the fact that you need to serialize the keys?

**Comment by tannerlinsley** [comment, marked as answer]
You can implement your own [Custom Query Key Serializer](https://github.com/tannerlinsley/react-query#custom-query-key-serializers-experimental) that does exactly this. It's definitely not the majority use-case, so like it says, there be dragons, but I know of others that have done this without any problems.

**Reply by pleunv** [reply]
Investigating react-query a bit as possible replacement for apollo and my first thought was also the same as @jackmellis'. The link above does not point to anything anymore and I can't immediately find it back in the docs, is this still possible today with v3?

edit: But I guess attempting to serialize a symbol is pretty pointless anyway since you might as well use a string then to guarantee uniqueness.

**Comment by kerematam** [comment]
I am visiting this issue as we just had a query key overlap in our project... and as I see there is no support of it.

I guess, the only option we got is our old friend: `Math.random()`

**Comment by TkDodo** [comment]
Sorry that the link to the accepted answer is not working anymore, I think this was a v2 thing.

Since v3, you can pass a `queryKeyHashFn` to `useQuery` or set it globally. It is documented [in the useQuery api docs](https://react-query.tanstack.com/reference/useQuery).

```
queryKeyHashFn: (queryKey: QueryKey) => string
Optional
If specified, this function is used to hash the queryKey to a string.
```

But since you need to produce a string to serialize a query key, Ithink this still stands:

> edit: But I guess attempting to serialize a symbol is pretty pointless anyway since you might as well use a string then to guarantee uniqueness.
